pkg/server/node.go

```diff
@@ -451,25 +453,32 @@ func (n *Node) start(
 		}
 		return nil
 	}); err != nil {
-		return errors.Wrapf(err, "failed to read last up timestamp from stores")
+		return errors.Wrapf(err, "failed to read last up timestamp from stores"), nil
 	}
 	n.lastUp = mostRecentTimestamp.WallTime
 
 	// Set the stores map as the gossip persistent storage, so that
 	// gossip can bootstrap using the most recently persisted set of
 	// node addresses.
 	if err := n.storeCfg.Gossip.SetStorage(n.stores); err != nil {
-		return fmt.Errorf("failed to initialize the gossip interface: %s", err)
+		return fmt.Errorf("failed to initialize the gossip interface: %s", err), nil
 	}
 
-	// Bootstrap any uninitialized stores.
-	//
-	// TODO(tbg): address https://github.com/cockroachdb/cockroach/issues/39415.
-	// Should be easy enough. Writing the test is probably most of the work.
+	// Bootstrap any uninitialized stores and define a function for blocking
+	// until new stores are fully bootstrapped. This function remains a
+	// no-op unless we find ourselves bootstrapping new stores.
+	waitForBootstrapNewStores = func() { }
 	if len(state.newEngines) > 0 {
-		if err := n.bootstrapStores(ctx, state.firstStoreID, state.newEngines, n.stopper); err != nil {
-			return err
-		}
+		// We need to bootstrap additional stores asynchronously because we
+		// can't rely on the KV layer to be available yet. See #39415.
```

Comment:
Let's use more words here. I like what we have in the commit message, so amending that, how about the following?

```
We need to bootstrap additional stores asynchronously. Consider the range that
houses the store ID allocator. When restarting the set of nodes that holds a
quorum of these replicas, when restarting them with additional stores, those
additional stores will require store IDs to get fully bootstrapped. But if we're
gating node start (specifically opening up the RPC floodgates) on having all
stores fully bootstrapped, we'll simply hang when trying to allocate store IDs.
See TestAddNewStoresToExistingNodes and #39415 for more details.

Instead we opt to bootstrap additional stores asynchronously, and rely on the
blocking function to signal to the caller that all stores have been fully
bootstrapped.
```

Feel free to include the same in the commit message as well.